shaders/EditorHexGrid.shader

```diff
@@ -1,27 +1,51 @@
 shader_type spatial;
 render_mode unshaded;
 
-// TODO: implement our own version of procedural hex grid tiling for a better
-// looking and more uniform hexes.
-// See andrewhungblog.wordpress.com/2018/07/28/shader-art-tutorial-hexagonal-grids/
+// Procedural hex grid tiling for a better looking and more uniform hexes.
+// This is heavily inspired by Andrew Hung's tutorial.
+// See andrewhungblog.wordpress.com/2018/07/28/shader-art-tutorial-hexagonal-grids
 
-uniform sampler2D gridTexture;
 uniform sampler2D maskTexture;
 uniform vec4 color : hint_color;
+uniform float lineWidth = 0.02;
+uniform float edgeLength = 1.3;
 
-uniform vec2 gridSize;
+const vec2 hexSize = vec2(1.7320508, 1.0); // 1.7320508 = sqrt(3)
+const vec2 hexSizeHalf = hexSize * 0.5;
 
 varying vec3 worldPos;
+varying float minLineWidth;
+
+// Returns normalized distance to the nearest hex center.
+float hexDist(vec2 coord)
+{
+    vec4 dists = mod(vec4(coord, coord - hexSizeHalf), hexSize.xyxy) - hexSizeHalf.xyxy;    
+    vec2 dist = dot(dists.xy, dists.xy) < dot(dists.zw, dists.zw) ? abs(dists.xy) : abs(dists.zw);
+    return max(dot(dist, hexSizeHalf), dist.y) + 0.5;
+}
 
 void vertex(){
-    worldPos = (WORLD_MATRIX * vec4(VERTEX, 1.0)).xyz;
+    worldPos = (CAMERA_MATRIX * vec4(0.0, 0.0, 0.0, 1.0)).xyz;
+
+    // min line width to not look unsteady
+    minLineWidth = (worldPos.y / VIEWPORT_SIZE.y);
 }
 
 void fragment(){
-    vec4 gridTex = texture(gridTexture, (worldPos.xz / gridSize) *
-        vec2(2.778, 4.811));
     vec4 mask = texture(maskTexture, UV);
 
-    ALBEDO = gridTex.rgb * color.rgb;
-    ALPHA = gridTex.a * mask.a;
+    if (mask.a == 0.0){
```

Comment:
Is this actually necessary? I see a multiply by mask.a in ALPHA later. I'm not a super shader expert so this might be incorrect, but I think in shaders if statements are way more expensive than doing unnecessary processing (as GPUs are not very efficient regarding branching).

Reply:
I thought so two, but found this [link](https://stackoverflow.com/a/30256689/2236081)
So I think this improves a bit.

Reply:
I guess the vast majority of pixel shader executions do early exit here, and the ones that don't are all grouped.